m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,745 @@
+            // Look for the first pair with an upper bound greater than or equal
+            // to the given upper bound.
+            if (bucketPairs[i].upperBoundDuration().compareTo(bucketUpperBound) >= 0) {
+                String idTagValue = String.format(bucketIdFmt, i);
+
+                tags.put(bucketIdTagName, idTagValue);
+                tags.put(bucketTagName,
+                    String.format("%s-%s",
+                        durationBucketString(bucketPairs[i].lowerBoundDuration()),
+                        durationBucketString(bucketPairs[i].upperBoundDuration())
+                    )
+                );
+
+                break;
+            }
+        }
+
+        reportCounterInternal(name, tags, samples);
+    }
+
+    // Relies on the calling function to provide guarantees of the reporter being open
+    private void reportCounterInternal(String name, Map<String, String> tags, long value) {
+        CountValue countValue = new CountValue();
+        countValue.setI64Value(value);
+
+        MetricValue metricValue = new MetricValue();
+        metricValue.setCount(countValue);
+
+        Metric metric = newMetric(name, tags, metricValue);
+
+        queueSizedMetric(new SizedMetric(metric, calculateSize(metric)));
+    }
+
+    private Metric newMetric(String name, Map<String, String> tags, MetricValue metricValue) {
+        Metric metric = new Metric(name);
+        metric.setTags(toMetricTagSet(tags));
+        metric.setTimestamp(System.currentTimeMillis() * Duration.NANOS_PER_MILLI);
+        metric.setMetricValue(metricValue);
+
+        return metric;
+    }
+
+    private void queueSizedMetric(SizedMetric sizedMetric) {
+        try {
+            metricQueue.put(sizedMetric);
+        } catch (InterruptedException e) {
+            LOG.warn(String.format("Interrupted queueing metric: %s", sizedMetric.getMetric().getName()));
+        }
+    }
+
+    private class Processor implements Runnable {
+        final Set<MetricTag> commonTags;
+        List<Metric> pendingMetrics = new ArrayList<>(freeBytes / 10);
+        int metricsSize = 0;
+
+        Processor(Set<MetricTag> commonTags) {
+            this.commonTags = commonTags;
+        }
+
+        @Override
+        public void run() {
+            try {
+                while (!executor.isShutdown()) {
+                    // This `poll` call will block for at most the specified duration to take an item
+                    // off the queue. If we get an item, we append it to the queue to be flushed,
+                    // otherwise we flush what we have so far.
+                    // When this reporter is closed, shutdownNow will be called on the executor,
+                    // which will interrupt this thread and proceed to the `InterruptedException`
+                    // catch block.
+                    SizedMetric sizedMetric = metricQueue.poll(MAX_PROCESSOR_WAIT_UNTIL_FLUSH_MILLIS, TimeUnit.MILLISECONDS);
```

Comment:
could you make this configurable, and change the default to 10s (default reporting interval in go-tally iirc)